Add service.UID into security group name

FengyunPan opened the pull request:
Related to: #53714 

**Release note**:
```release-note
NONE
```

anguslees reviewed (commented):
This lgtm.  We also need to handle the migration from old securityGroup name -> new.  Shall we add that in this PR or another followup?

FengyunPan commented:
@anguslees Sorry, I don't know why we need to look up the old name during update/delete (not create).
The servie controller just calling UpdateLoadBalancer() when the nodes of cluster are changed, service controller can not update the name of service and the name of security group.
Can you tell more if my understand is wrong?

FengyunPan commented:
/test pull-kubernetes-unit

anguslees commented:
> Sorry, I don't know why we need to handle the migration from old securityGroup name -> new.( look up the old name during update/delete (not create))

If someone has an existing type=LoadBalancer, upgrades kubernetes to this new provider code and then does something that triggers one of the `*LoadBalancer()` functions - what will happen?

I think `CreateLB()` obviously works ok (no history to worry about).  I think `EnsureLB()` will decide the new securityGroup is missing and create a new one - but I think the existing host securityGroup will still refer to the old/stale lbsg (? haven't actually read the code in detail).  I think both `EnsureLB()` and `EnsureLBDeleted()` will leak the old securityGroup and never remove it.

So: I think we need to do some sort of extra work to handle the transition between names smoothly.  I think the easiest way to do this is to look for the new name near the top of the main entry points (`EnsureLB*`) and if we don't find it, then look for the old name, and if we *do* find the old name, then rename the existing sg to the new name (assuming we can do a rename in openstack?).  This ensures references from other securitygroups still work, and means the rest of the code can just assume only the new name is relevant.

FengyunPan commented:
> I think EnsureLB() will decide the new securityGroup is missing and create a new one - but I think the existing host securityGroup will still refer to the old/stale lbsg 

I don't think the old name and the new name will be exist. EnsureLB() is used to ensure LB resource and ensureSecurityGroup() is used to ensure security group. If service's security group is exist, ensureSecurityGroup() just check the status of security group and don't create a new security group.
So the getSecurityGroupName() is ok, like cloudprovider.GetLoadBalancerName().

dims commented:
/retest

FengyunPan commented:
@anguslees I think I can't understand the issue which in your comment, so this PR just fix "unique".
Is it need to file a new issue for your issuse?

anguslees commented:
So this change fixes the securityGroup name to be unique across time, good.

My other issue is how we handle the _migration_ from the old name to the new name.  Consider this timeline:
1. Cluster deployed using older k8s.  Lots of type=LoadBalancer Services, life is happy.  All the LB SGs use the older names.
2. Cluster controller manager is upgraded to newer k8s, which includes #53714.
3. A user changes or deletes one of those older type=LoadBalancer Services.
4. `EnsureLB`/`DeleteLB` will not find the SG under the new name, and will create a new SG, or ignore the fact (for `DeleteLB`).
5. The new entry will work, but the SG with the old name (pre-#53714) will be leaked and never cleaned up.

Perhaps more seriously, the existing node security group will continue to refer to the old LB SG name, so won't respond to changes in the LB SG correctly (although I'm not sure what triggers such a change.  Perhaps LB replacement?).

As I've described elsewhere, the fix is just to look for the old name, and rename the existing SG to the new name at the right point(s).

I have no opinion on whether the transition logic should be added in a different PR or not, but it's only a few lines of code, and it needs to be fixed.  (Ideally we would go back in time and ensure the original PR handled these correctly before merging)

FengyunPan commented:
@anguslees Oh, I see. Add codes for it soon.

FengyunPan2 committed:
Add service.UID into security group name

Related to: #53714

FengyunPan commented:
/retest

FengyunPan commented:
@anguslees @dims Any thoughts?

dims commented:
@FengyunPan i'll wait for @anguslees 's input. in general it looks good

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -868,6 +874,14 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+		// TODO(FengyunPan): Remove it at V1.10
+		err = lbaas.EnsureOldSecurityGroupDeleted(clusterName, apiService)
+		if err != nil {
+			return status, fmt.Errorf("Failed to delete the Securty Group for loadbalancer service %s/%s: %v", apiService.Namespace, apiService.Name, err)
```
Typo: *Security*

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -914,8 +928,8 @@ func (lbaas *LbaasV2) ensureSecurityGroup(clusterName string, apiService *v1.Ser
-			Name:        getSecurityGroupName(clusterName, apiService),
-			Description: fmt.Sprintf("Securty Group for loadbalancer service %s/%s", apiService.Namespace, apiService.Name),
+			Name:        getSecurityGroupName(apiService),
+			Description: fmt.Sprintf("Securty Group for %s/%s Service LoadBalancer in cluster %s", apiService.Namespace, apiService.Name, clusterName),
```
Again: *Security*

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1368,50 +1382,131 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
-			} else {
-				return fmt.Errorf("error occurred finding security group: %s: %v", lbSecGroupName, err)
-			}
+			return fmt.Errorf("Failed to delete Securty Group for loadbalancer service %s/%s: %v", service.Namespace, service.Name, err)
```
It's everywhere!  Typo: *Security*

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1368,50 +1382,131 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
+		// TODO(FengyunPan): Remove it at V1.10
+		err = lbaas.EnsureOldSecurityGroupDeleted(clusterName, service)
+		if err != nil {
+			return fmt.Errorf("Failed to delete the Securty Group for loadbalancer service %s/%s: %v", service.Namespace, service.Name, err)
```
*Security*

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1368,50 +1382,131 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
+		// Just happen when nodes have not Security Group, or should not happen
+		// UpdateLoadBalancer and EnsureLoadBalancer can set lbaas.opts.NodeSecurityGroupIDs when it is empty
+		// And service controller call UpdateLoadBalancer to set lbaas.opts.NodeSecurityGroupIDs when controller manager service is restarted.
+		glog.Warningf("Can not find node-security-group from all the nodes of this cluser when delete loadbalancer service %s/%s",
```
Typo: *cluster*

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1368,50 +1382,131 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
+		// Just happen when nodes have not Security Group, or should not happen
+		// UpdateLoadBalancer and EnsureLoadBalancer can set lbaas.opts.NodeSecurityGroupIDs when it is empty
+		// And service controller call UpdateLoadBalancer to set lbaas.opts.NodeSecurityGroupIDs when controller manager service is restarted.
+		glog.Warningf("Can not find node-security-group from all the nodes of this cluser when delete loadbalancer service %s/%s",
```
Typo: *cluster*

anguslees reviewed (changes requested):
lgtm, but there's some super-minor typos in error messages that we may as well fix now before merging.

FengyunPan2 committed:
Add EnsureOldSecurityGroupDeleted to delete old security group

Consider the migration from the old security group name to the new
security group name, we need delete the old security group.
At V1.10, we can assume everyone is using the new security group
names and remove this code.

FengyunPan commented:
Oops, fixed them.

FengyunPan commented:
/test pull-kubernetes-e2e-gce

anguslees commented:
/approve

FengyunPan commented:
@jdumars @calebamiles Can you please add the 1.9 milestone flag for this PR? Thank you.

anguslees commented:
Adding some labels so this doesn't slip through the cracks during 1.9 release shenanigans.  We want this in 1.9 so deployments don't have to do the LB name change twice (and the first time without migration logic).

/sig openstack
/kind bug
/priority critical-urgent

dims commented:
@enisoc requesting on behalf of sig-openstack, please add this to 1.9 milestone

cc @idvoretskyi @xsgordon

dims commented:
/sig openstack

FengyunPan commented:
@dims Can you please add 'lgtm' label? Thank you.

dims commented:
/lgtm

of course. sorry
